cudnn: 7.6.5 -> 8.3.2

samuela opened the pull request:
###### Description of changes
Note that this PR is targeting staging instead of master because the PyTorch version on staging (v1.11.0) is more amenable to cudnn upgrades than the PyTorch version on master (v1.10.2).

This PR also addresses https://github.com/NixOS/nixpkgs/pull/166359#issuecomment-1082737458 to follow up on my sworn promise to rid nixpkgs of cudnn_8_1_cudatoolkit_11.

Running tools like nixpkgs-review on this PR is going to be tricky atm due to https://github.com/NixOS/nixpkgs/issues/166444, which thus far we are led to believe is a temporary CDN interruption. However, I have built pytorchWithCuda successfully since I already had cudnn builds cached. 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

samuela committed:
cudnn: formatting

samuela committed:
cudnn: 7.6.5 -> 8.3.2

SomeoneSerge commented on pkgs/development/libraries/science/math/cudnn/default.nix:
```diff
@@ -81,31 +81,27 @@ rec {
   cudnn_8_1_cudatoolkit_11_2 = cudnn_8_1_cudatoolkit_10_2.override { cudatoolkit = cudatoolkit_11_2; };
 
   cudnn_8_1_cudatoolkit_10 = cudnn_8_1_cudatoolkit_10_2.override { cudatoolkit = cudatoolkit_10; };
-  cudnn_8_1_cudatoolkit_11 = cudnn_8_1_cudatoolkit_10_2.override { cudatoolkit = cudatoolkit_11; };
```
Deleted from `cudnn/default.nix` but, apparently, still `inherit`-ed in `all-packages.nix`: https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree/jobs/101

samuela replied:
Ah good catch! fixed in https://github.com/NixOS/nixpkgs/pull/166533/commits/5e686a3a78df40f06b0c1dc064b130fe4a9d9318

samuela committed:
cudatoolkit_11: 11.6.1 -> 11.5.0

samuela committed:
cutensor: fix cutensor_cudatoolkit_{10,11}

samuela committed:
cupy: check that all cudatoolkit versions are the same

samuela committed:
cudnn_8_1_cudatoolkit_11: remove

samuela commented:
Notes:
* I uncovered an issue with cupy where it would pull in both cudatoolkit_11_5 and cudatoolkit_11_4. This was bc `cutensor_cudatoolkit_11 = cutensor_cudatoolkit_11_4;` instead of overriding with cudatoolkit_11. I fixed this in https://github.com/NixOS/nixpkgs/pull/166533/commits/d7f9d3a6db3968b26a240ca3b6025663eb1cf7db. To avoid this failure in the future I added asserts in cupy/default.nix to double check that we're using consistent cudatoolkit versions across all of our dependencies.
* cudatoolkit_11 was pointing to cudatoolkit_11_6 due to https://github.com/NixOS/nixpkgs/pull/165021. This has already been fixed on master (https://github.com/NixOS/nixpkgs/pull/166359), but not on staging. I fixed inhttps://github.com/NixOS/nixpkgs/pull/166533/commits/259ac51d2f66425880ff094822dee6a01f4acab1 and added a comment noting the dangers of updating cudatoolkit_11.
* cutensor is currently at 1.3.1, but the latest version is 1.5.0. Refraining from updating in this PR.
* cupy is currently at 10.2.0, but the latest version is 10.3.0. Refraining from updating in this PR.

samuela committed:
Merge branch 'staging' into samuela/cudnn2

samuela commented:
I'm going to close this in favor of https://github.com/NixOS/nixpkgs/pull/166784 which is (roughly) the same thing but targeting master instead of staging. If that fails for some reason, we can revive this PR.
